### PR TITLE
feat(s3): website routing rules

### LINKED ISSUE #3409: S3: Missing websiteRoutingRules properties
**Note: for support questions, please first reference our [documentation](https://docs.aws.amazon.com/cdk/api/latest), then use [Stackoverflow](https://stackoverflow.com/questions/ask?tags=aws-cdk)**. This repository's issues are intended for feature requests and bug reports.

* **I'm submitting a ...**
  - [ ] :beetle: bug report
  - [ ] :rocket: feature request
  - [x] :books: construct library gap
  - [ ] :phone: security issue or vulnerability => Please see [policy](https://github.com/aws/aws-cdk/security/policy)
  - [ ] :question: support request => Please see note at the top of this template.

* **What is the current behavior?**
*If the current behavior is a :beetle:bug:beetle:: Please provide the steps to reproduce*
The [WebsiteConfiguration RoutingRules](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-websiteconfiguration.html#cfn-s3-websiteconfiguration-routingrules) is currently missing from the S3 package

* **What is the expected behavior (or behavior of feature suggested)?**
Expose a `websiteRoutingRules` property

* **What is the motivation / use case for changing the behavior or adding this feature?**
Add missing L2 poperty

* **Please tell us about your environment:**

  - CDK CLI Version: 1.1.0
  - Module Version: 1.1.0
  - OS: all
  - Language: all

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. associated pull-request, stackoverflow, gitter, etc)

### PR DESCRIPTION
**Pull request by nmussy**
Fixes #3409

----

Please read the [contribution guidelines](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md) and follow the pull-request checklist.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by nmussy** [comment]
The [API reference](https://docs.aws.amazon.com/AmazonS3/latest/API/API_Redirect.html) mentions a few times: 
> Not required if one of the siblings is present

From what I've tested, this is actually never true. S3 accepts every silly rule you can give it, e.g.:

```xml
<!-- redirects to /testtesttesttest... until the browser stops following them-->
<RoutingRules><RoutingRule><Redirect>
	<ReplaceKeyPrefixWith>test</ReplaceKeyPrefixWith>
</Redirect></RoutingRule></RoutingRules>
```

Most of these result in an infinite redirection. I don't think it's the CDK's responsibility to catch those, do you agree?

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 1]
```diff
@@ -1485,6 +1504,106 @@ export enum BucketAccessControl {
...
+   */
+  readonly condition?: RoutingRuleCondition;
+}
+
+/**
+ * Rule that define when a redirect is applied and the redirect behavior.
+ *
+ * @see https://docs.aws.amazon.com/AmazonS3/latest/dev/how-to-page-redirect.html
+ */
+export class RoutingRule {
```
Comment (eladb) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts]: Why do we need this as a class? Just rename `RoutingRuleProps` to `RoutingRule` and move the validation code to the main construct

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 2]
```diff
@@ -1485,6 +1504,106 @@ export enum BucketAccessControl {
...
+  readonly protocol?: RedirectProtocol;
+
+  /**
+   * Specifies the object key prefix to use in the redirect request
+   *
+   * Not required if one of the siblings is present.
+   *
+   * @default - The key will not be replaced
+   */
+  readonly replaceKey?: ReplaceKeyWith | ReplaceKeyPrefixWith;
```
Comment (eladb) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts]: Unions are not allowed. Use "union-like" classes:

```ts
class ReplaceKey {
  public static with(...);
  public static prefixWith(...);
```

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 3]
```diff
@@ -1485,6 +1504,106 @@ export enum BucketAccessControl {
...
+   * The host name to use in the redirect request
+   *
+   * @default - The host name used in the original request.
+   */
+  readonly hostName?: string;
+
+  /**
+   * The HTTP redirect code to use on the response
+   *
+   * Not required if one of the siblings is present.
```
Comment (eladb) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts]: What does that mean?

Comment (nmussy) [reply packages/@aws-cdk/aws-s3/lib/bucket.ts]: Pretty much [nothing](https://github.com/aws/aws-cdk/pull/3411#issuecomment-514557933), I'll remove it

File: packages/@aws-cdk/aws-s3/lib/bucket.ts (hunk 1 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-s3/lib/bucket.ts]: Done

File: packages/@aws-cdk/aws-s3/lib/bucket.ts (hunk 2 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-s3/lib/bucket.ts]: Done

File: packages/@aws-cdk/aws-s3/lib/bucket.ts (hunk 3 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-s3/lib/bucket.ts]: Done

**Comment by eladb** [comment]
@nmussy could you please add a README section about this new feature and we'll merge it in?

**Comment by nmussy** [comment]
@eladb Done. I preceded it with a `websiteRedirect` section

**Pull request merged**